OpenSim/Simulation/OpenSense/ModelCalibrator.cpp

```diff
@@ -0,0 +1,234 @@
+//_____________________________________________________________________________
+/**
+ * Default constructor.
+ */
+ModelCalibrator::ModelCalibrator() 
+{
+    constructProperties();
+}
+
+
+ModelCalibrator::ModelCalibrator(const std::string& setupFile)
+        : Object(setupFile, true) {
+    constructProperties();
+    updateFromXMLDocument();
+}
+
+//_____________________________________________________________________________
+/**
+ * Destructor.
+ */
+ModelCalibrator::~ModelCalibrator()
+{
+}
+
+//=============================================================================
+// CONSTRUCTION
+
+//_____________________________________________________________________________
+/**
+ */
+void ModelCalibrator::constructProperties() 
+{
+    constructProperty_model_file_name("");
+    constructProperty_base_imu_label("z");
+    constructProperty_base_heading_axis("pelvis_imu");
+    constructProperty_sensor_to_opensim_rotations(
+            SimTK::Vec3(-SimTK_PI / 2, 0, 0));
+    constructProperty_calibration_file_name("");
+}
+
+//=============================================================================
+// UTILITY
+//=============================================================================
+//_____________________________________________________________________________
+/**
+ * This method runs the calibration method on the _model maintained by
+ * this ModelCalibrator
+ */
+bool ModelCalibrator::run(bool visualizeResults)  {
+    if (_model.empty()) { 
+        _model.reset(new Model(get_model_file_name())); 
+    }
+    TimeSeriesTable_<SimTK::Quaternion> quatTable(get_calibration_file_name());
+
+    const SimTK::Vec3& sensor_to_opensim_rotations = get_sensor_to_opensim_rotations();
+    SimTK::Rotation sensorToOpenSim =
+            SimTK::Rotation(SimTK::BodyOrSpaceType::SpaceRotationSequence,
+                    sensor_to_opensim_rotations[0], SimTK::XAxis,
+                    sensor_to_opensim_rotations[1], SimTK::YAxis,
+                    sensor_to_opensim_rotations[2], SimTK::ZAxis);
+    // Rotate data so Y-Axis is up
+    OpenSenseUtilities::rotateOrientationTable(quatTable, sensorToOpenSim);
+
+    std::string imu_axis = IO::Lowercase(get_base_heading_axis());
+    SimTK::CoordinateDirection directionOnIMU(SimTK::ZAxis);
+    int direction = 1;
+    if (imu_axis.front() == '-') 
+        direction = -1;
+    char& back = imu_axis.back();
+    if (back == 'x')
+        directionOnIMU = SimTK::CoordinateDirection(SimTK::XAxis, direction);
+    else if (back == 'y')
+        directionOnIMU = SimTK::CoordinateDirection(SimTK::YAxis, direction);
+    else if (back == 'z')
+        directionOnIMU = SimTK::CoordinateDirection(SimTK::ZAxis, direction);
+    else { // Throw, invalid specification
+    
+    }
+
+    // Compute rotation matrix so that (e.g. "pelvis_imu"+ SimTK::ZAxis) lines up
+    // with model forward (+X)
+    SimTK::Rotation headingRotation =
```

Comment:
Could we return the angularDifference from OpenSenseUtilities::computeHeadingCorrection and make a get() method from the model calibrator?

When debugging the orientations, it is helpful to know what the angular difference is but it is printed to the console above all of the imu frame orientations so you have to scroll up and try and find it. It would be nice if the value could be gettable (.getAngluarDifference() ) or printable (.printAngluarDifferenceToConsol() ).

Reply:
Good ideas @jimmyDunne I was debating whether computeHeadingCorrection would be better buried inside the ModelCalibrator as the only client or keep at OpenSenseUtilities for potential use by other calibration schemes. If we move it inside ModelCalibrator then it can be maintained as a part of the ModelCalibrator object, retrieving the angle and/or the Transform. Any thoughts?

Reply:
From chatting offline about this, it feels like the thing to do would be to keep computeHeadingCorrection() in OpenSenseUtilities so other people could use it if needed. The function could return the angle and the Rotation matrix could be constructed in ModelCalibrator().

Reply:
@jimmyDunne to strike a balance between returning a Rotation matrix vs returning a double with the implicit assumption that it is represents rotation about Y, I modified the method to return  a Vec3 of rotations that the caller can convert to a Rotation. Please let me know what you think.